torch/_dynamo/variables/higher_order_ops.py

```diff
@@ -1029,64 +1109,70 @@ def call_function(
             )
         additional_inputs_seq = additional_inputs.unpack_var_sequence(tx)
 
-        (
-            (cond_r, cond_treespec),
-            cond_graph,
-            cond_lifted_freevars,
-        ) = speculate_subgraph(
-            tx,
-            cond_fn,
-            operands_seq + additional_inputs_seq,
-            {},
-            "while_loop",
-            source_target=self.value,
-            set_subgraph_inputs="manual",
-        )
-        cond_nn_modules = dict(tx.output.nn_modules)
-        if not isinstance(cond_r, TensorVariable):
-            unimplemented(
-                f"Expected cond_fn to return a tensor but got {cond_r.python_type()}",
+        with self._temp_lift_int_args_as_unbacked_symint_inputs(
+            tx, operands_seq
+        ) as subgraph_carries:
+            (
+                (cond_r, cond_treespec),
+                cond_graph,
+                cond_lifted_freevars,
+            ) = speculate_subgraph(
```

Comment:
The way I would have implemented this is:
- for each int in operands_seq, create new unbacked symints
- call speculate subgraph on the the new operands_seq.

The thing that confuses me are:
- why does this API need to be a context manager?
- why do we need to create dummy placeholders for the unbacked symint?

Reply:
> why do we need to create dummy placeholders for the unbacked symint?

This is to workaround the invariant that each SymNode variable need to have a proxy. I created a placeholder for this purpose. I guess I could create a call function proxy (e.g. create_unbacked_symint) and remove it after finish speculating the subgraph but it's essentially the same thing.

> why does this API need to be a context manager?
we want to delete the dummy placeholder after speculation.

Reply:
flexattention also has a problem where it just creates random dummy proxies. Can we have some sort of generic helper function for temporarily creating proxies and then deleting them?

Since that seems to be where we're going in general with HOP tracing ("temporarily create new proxies, trace with them, and then throw them out") that should probably be the main API rather than adding new placeholders

Reply:
yeah, that's true. 

Let me think about this. One approach is to do this automatically in speculate_subgraph but we need to design the interface to allow customizations (e.g. we don't always want to cast int to unbacked symint)

Reply:
we probably also do this in map, right? We need to slice the inputs so that we can speculate_subgraph on the sliced inputs

Reply:
Yeah, scan/associative_scan also works on a slice of inputs. basically, any hops that's not wrap-like, might potentially need to create temp variables.

So I feel concretely the speculate_subgraph should take **exmaple_values**  (e.g. fak tensors, symints) instead of variable trackers (kind of like the builder.wrap). Because what we do right now in validate_args_and_maybe_create_graph_inputs is just extracting "exmaple_value" and create graph_input using the example_value.

Reply:
oh, we can't we because we need the variable tracker to call " f.call_function(tx, args, sub_kwargs)"